src/common/tools/semver.cpp

```diff
@@ -0,0 +1,434 @@
+    Error  err;
+
+    while (true) {
+        Tie(identPos, err) = version.FindSubstr(identPos, ".");
+        if (err.IsNone()) {
+            version[identPos] = '\0';
+            identPos++;
+        }
+
+        if (auto pushErr = identifiers.PushBack(pos); !pushErr.IsNone()) {
+            return pushErr;
+        }
+
+        if (!err.IsNone()) {
+            if (!err.Is(ErrorEnum::eNotFound)) {
+                return err;
+            } else {
+                return ErrorEnum::eNone;
+            }
+        }
+
+        pos = version.begin() + identPos;
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ValidateNumericIdentifier(const String& identifier)
+{
+    if (auto err = ValidateChars(identifier, cNumericChars); !err.IsNone()) {
+        return err;
+    }
+
+    if (identifier.Size() == 0) {
+        return ErrorEnum::eInvalidArgument;
+    }
+
+    if (identifier.Size() > 1 && identifier[0] == '0') {
+        return ErrorEnum::eInvalidArgument;
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ValidateStrIdentifier(const String& identifier)
+{
+    if (auto err = ValidateChars(identifier, cAllowedChars); !err.IsNone()) {
+        return err;
+    }
+
+    if (identifier.Size() == 0) {
+        return ErrorEnum::eInvalidArgument;
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ValidateBasePart(String& version)
+{
+    StaticArray<String, cMaxNumIdentifiers> identifiers;
+
+    if (auto err = GetIdentifiers(version, identifiers); !err.IsNone()) {
+        return err;
+    }
+
+    if (identifiers.Size() == 0 || identifiers.Size() > 3) {
+        return ErrorEnum::eInvalidArgument;
+    }
+
+    for (const auto& identifier : identifiers) {
+        if (auto err = ValidateNumericIdentifier(identifier); !err.IsNone()) {
+            return err;
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ValidatePrereleasePart(String& version)
+{
+    StaticArray<String, cMaxNumIdentifiers> identifiers;
+
+    if (auto err = GetIdentifiers(version, identifiers); !err.IsNone()) {
+        return err;
+    }
+
+    for (const auto& identifier : identifiers) {
+        if (IsNumericIdentifier(identifier)) {
+            if (auto err = ValidateNumericIdentifier(identifier); !err.IsNone()) {
+                return err;
+            }
+        } else {
+            if (auto err = ValidateStrIdentifier(identifier); !err.IsNone()) {
+                return err;
+            }
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ValidateMetadataPart(String& version)
+{
+    StaticArray<String, cMaxNumIdentifiers> identifiers;
+
+    if (auto err = GetIdentifiers(version, identifiers); !err.IsNone()) {
+        return err;
+    }
+
+    for (const auto& identifier : identifiers) {
+        if (auto err = ValidateStrIdentifier(identifier); !err.IsNone()) {
+            return err;
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error SplitVersion(String& version, Array<String>& parts)
```

Comment:
it looks the array consists of 3 elements in all the cases. 
mb better split it into 3 parameters